client/core/trade.go

```diff
@@ -1468,51 +1554,94 @@ func (c *Core) tick(t *trackedTrade) (assetMap, error) {
 			return ctx.Err()
 		}
 
-		if t.shouldBeginFindRedemption(ctx, match) {
+		// A redemption search should only be done for a revoked match, but
+		// MatchProof may not be updated, so we enforce that here.
+		if revoked && t.shouldBeginFindRedemption(ctx, match) {
 			c.log.Debugf("Ready to find counter-party redemption for match %s, order %v (%v)", match, t.ID(), side)
-			t.findMakersRedemption(c.ctx, match) // async search, not on the timeout
+			searches = append(searches, match)
 			return nil
 		}
 
 		return ctx.Err()
 	}
 
+	// Begin checks under read-only lock.
+	t.mtx.RLock()
+	// Check all matches for and resend pending requests as necessary.
+	c.resendPendingRequests(t)
+
 	// Check all matches and send swap, redeem, or refund as necessary.
 	for _, match := range t.matches {
-		if err := checkMatch(match); err != nil {
-			return assets, err
+		err := checkMatch(match)
+		if err == nil {
+			continue
 		}
+		// With a timeout or shutdown, all this trade's matches will have same
+		// fate, so return now.
+
+		t.mtx.RUnlock()
+
+		if len(revokes) != 0 {
+			// Still flag any "should revoke"s for IsRevoked() and to fast track
+			// the next tick. NOTE: See the TODO below regarding revokeMatch.
+			t.mtx.Lock()
+			defer t.mtx.Unlock()
+			for _, rm := range revokes {
+				rm.MetaData.Proof.SelfRevoked = true
+			}
+		}
+
+		return assets, err
 	}
+	// End checks under read-only lock.
+	t.mtx.RUnlock()
 
-	fromID := t.wallets.fromAsset.ID
 	if len(swaps) > 0 || len(refunds) > 0 {
-		assets.count(fromID)
+		assets.count(t.wallets.fromAsset.ID)
 	}
+	if len(redeems) > 0 {
+		assets.count(t.wallets.toAsset.ID)
+		assets.count(t.fromAssetID) // update ContractLocked balance
+	}
+
+	if len(swaps) == 0 && len(refunds) == 0 && len(redeems) == 0 &&
+		len(revokes) == 0 && len(searches) == 0 {
+		return assets, nil // nothing to do, don't acquire the write-lock
+	}
```

Comment:
nit: Could also do `if len(swaps)|len(refunds)|len(redeems)|len(revokes)|len(searches) == 0 {`

Reply:
True, and that's a nice trick, and same result, but with no short-circuiting.  The length of every slice need not be checked here if, say, `swap` were not empty.  NBD either way.